### ISSUE TITLE
[DISCUSSION] Layer-by-Layer Graph Execution

### ISSUE THREAD

**Issue by rjzamora** [dask/dask#8616]
### TLDR

**Rough Proposal**: In order to overlap graph communication with scheduling and computation, we should enable **partial** task-graph execution. Further details are discussed below, but the ultimate goals is to completely remove the need to send HLGs to the scheduler (while hopefully improving Dask's ability to handle large-scale graphs).

**NOTE:** The ideas here are mostly a combination of suggestions made in the past by others like @madsbk, @gjoseph92, @ian-r-rose (and probably others)

## Background

Most methods in the Dask-Dataframe and Dask-Array APIs produce an abstract `HighLevelGraph` (HLG) instead of a fully-materialized dictionary-based task graph. The general purpose of this HLG approach is two-fold:

1. To enable critical graph-optimization operations **before** the full (dictionary-based) task graph is materialized. These optimizations include blockwise task fusion, culling, and partial-IO optimizations like column projection (and hopefully predicate pushdown in the near future).
2. To avoid sending large materialized task graphs over the wire from the client process to the scheduler. We expected (and indeed observe) much smaller graph-communication latency when the HLG-to-dictionary materialization was moved from the client to the scheduler.

Although HLGs have demonstrated some success in achieving both of the above goals, there have also been clear challenges:

- HLG fusion has relied on existing optimizations in `Blockwise`, requiring all “fusible” DataFrame and Array Layers (including IO) to be reformulated as `Blockwise`-based HLG Layers.  The complexity in `Blockwise` has made this transition a bit difficult.  I will not focus on this specific challenge in the remainder of this proposal (but I felt it was important enough to mention).
- The need to deserialize and materialize HLGs on the scheduler has proved to be a complete headache (see [distributed#5581](https://github.com/dask/distributed/issues/5581) and [this summary](https://github.com/dask/dask/pull/8468#discussion_r771752382)). The existing serialization and communication infrastructure was designed with the low-level (dictionary) graph in mind.  We quickly discovered that it can be quite painful to construct a low-level graph on the scheduler without being able to call `pickle.loads` on functions and non-msgpack serializable arguments (a security concern).
- Sending HLGs to the scheduler reduced the graph communication time (important for high-latency client-scheduler connections), but it barely made a dent in typical end-to-end run times for systems with decent client-scheduler latency. This is because large client-side graph materialization costs were simply converted into comparable scheduler-side graph materialization costs.  We originally hoped that Cython or C++ on the scheduler would make this graph-materialization cost negligible, but I am skeptical that this is something we can/should rely on.

What I take away from the challenges above is that: (1) Sending HLGs to the scheduler is a headache, and (2) The only reason we are still doing this is to minimize client-to-scheduler communication latency.

## Proposal

After a long (but fun) discussion with @madsbk, we generally agreed that we may be able to avoid sending HLGs to the scheduler altogether. Our primary (and probably lone) reason for dealing with HLGs on the scheduler now is graph-communication latency.  Therefore, we will no longer have any motivation to send the HLG to the scheduler if this overhead is sufficiently reduced or hidden.  One possible way to do this is by overlapping graph communication with scheduling and execution.  That is, we may be able to graph communication delays by taking the following steps:

### Step 1 - Enable partial graph scheduling

The first and most-critical step is to ensure that the scheduler is capable of executing a partial graph.  I am hoping someone like @madsbk will have the bandwidth and interest to prototype this machinery, but I would expect the client to be able to submit a graph without specifying the keys it wants to `get`.  Instead, the client should be able to mark specific keys as active dependencies.  Note that I am very open to suggestions here, but the overall goal is to begin assigning tasks to workers before the scheduler has finished receiving the full task graph.

### Step 2 - Send the materialized graph Layer-by-Layer

Once the scheduler has a mechanism for processing partial task-graphs, we will want to add a mechanism in the client to actually send large graphs in phases. For the initial pass, I propose that we simply use the existing HLG `Layer` as a single “sub-graph” unit.  In other words, after the HLG optimization pass, we can simply send the materialized graph to the scheduler one `Layer` at a time.  After we confirm that this “works,” we can avoid the obvious small-message problem by aggregating adjacent `Layers` on the client until an optimal sub-graph size is reached. 

If we are able to implement these first two steps without critical road blocks, then I suspect we will be highly motivated to discuss the future (or lack thereof) of HLG serialization.  If we can decouple the serialization design from that of HLGs, then I suspect we will find it much easier to improve the two separately.

## Possible Follow-On Work

Although Layer-by-Layer graph communication/execution may largely reduce our need to send HLGs to the scheduler, we may still want to send something other than the low-level graph is “special cases.”  For example, if the user is willing to **opt in** to “full” (`dumps` and `loads`) pickle support on the scheduler, then we may want to enable the client to pickle a single generator function for an HLG Layer, and the scheduler could simply unpickle this object and then pass in a required set of output keys to get the necessary tasks.  This would allow the scheduler to break up the materialization and scheduling of extremely large graphs (even when a single layer contains billions of keys).  Such an optimization would clearly take considerable design and development effort, but it may provide us with a reasonable path toward the handling of “extreme-scale” graphs (note that this idea belongs entirely to @madsbk, so he makes want to clarify and/or expand)

**Comment by rjzamora** [comment]
cc @quasiben @jakirkham @mrocklin @jcrist (critical thoughts very welcome)

**Comment by gjoseph92** [comment]
@rjzamora thanks for writing this up; I've been wanting to discuss something similar.

Layer-by-layer will be hard to schedule well. It might make root task overproduction even worse (https://github.com/dask/distributed/issues/5555 / https://github.com/dask/distributed/issues/5223). Since layers are essentially breadth-first, and execution order should be essentially depth-first, I don't think we'd actually be able to overlap the computation with the communication very well. Or rather, if we did so, we'd be doing lots of the _wrong_ computations while waiting for the next communication to happen. Or, if we did so smartly, we'd basically say "okay, I got this root layer of 100,000 tasks, but I'm only scheduling 100 of them and then waiting to get the next layer"—in which case, you're not getting much benefit from overlapping computation and communication, since those 100 may well finish before the next layer is even received.

OTOH, we could look at streaming the full graph to the scheduler in priority order—basically, allow computation to start before all of the tasks have been received. This sounds nice, until you realize that it still requires materializing the full graph on the client (so you can prioritize it) before sending anything to the scheduler. So you're still not getting much benefit there with large graphs.

-----

What I've been thinking about instead is whether we can avoid materializing the graph at all—even on the scheduler—by leveraging the high-level structure of the graph for scheduling purposes.

**Basically, could we define an interface on HLGs where you could iterate through tasks in the HLG in approximate priority order ("good scheduling order", aka depth-first order) without materializing the layers?**

This way, you could have an HLG of arbitrary size (billions of tasks), but you'd hopefully only have about `nthreads` tasks actually in memory on the scheduler at once.

**Comment by rjzamora** [comment]
>Basically, could we define an interface on HLGs where you could iterate through tasks in the HLG in approximate priority order ("good scheduling order", aka depth-first order) without materializing the layers?

This idea may be similar to what Mads had in mind, and what I poorly described at the end of the issue:  Perhaps we just want to be able to pickle some kind of generator function that the scheduler can simply unpickle and use to generate tasks in a reasonable order (not necessarily by layer, but maybe by required output key).  However, in order to get something like this working without a serialization nightmare, it seems like the user would need to opt in to "unpickling on the scheduler". The reason I'm suggesting partial graph execution in general, is so distributed exectution could still "work" for users who do not want "unpickling on the scheduler". In that case, the user would use the same generator function on the client side to materialize subgraphs and send along to the scheduler.

I completely agree that a layer-by-layer approach may not lead to optimal scheduling. In the long run we definitely want to revise the HLG (or perhaps high-level expression) system to generate tasks is a more "functional" way. My suggestions to send the graph is Layer batches is mostly as a path to prototype the partial-exectution idea with minimal changes. However, if you think that the graph-generator logic should come first, I am also open to that idea :)

**Comment by gjoseph92** [comment]
> Perhaps we just want to be able to pickle some kind of generator function that the scheduler can simply unpickle and use to generate tasks in a reasonable order (not necessarily by layer, but maybe by required output key)

I think we're circling around the idea that you, me, @ian-r-rose, and @GenevieveBuckley talked about. With low-level graphs, you figure out the outputs given the inputs. For high-level graphs to be effective, you need to work backwards: you must figure out the inputs given the outputs. You have to invert all the graph-generating logic currently in dask/dask, often coming up with new algorithms. It's a lot of work—it's not just wrapping existing code in a new interface. But if you don't do this, there's not much benefit to high-level graphs, since you're just deferring the materialization to some other point in time.

In order to make this iterate-in-priority-order concept work—which is where I think we'd _really_ see HLGs pay off—we'd have to go all-in on flipping these graph-generation algorithms on their heads.

**Comment by gjoseph92** [comment]
> My suggestions to send the graph is Layer batches is mostly as a path to prototype the partial-exectution idea with minimal changes

I like the incremental approach, I just don't see how much value it would add. Worst-case, it would make an existing problem (root-task execution) even worse. But best-case, communication and computation just wouldn't overlap much, so you wouldn't gain much performance. I just don't think there's much of a way to work around needing depth-first scheduling, but having breadth-first communication.

That said, I'd love to be able to stream the transmission and especially unpacking of HLGs, even if we don't start executing them until the whole thing is materialized. Currently, as you mentioned, materializing the HLG (or deserializing and processing a low-level graph) on the scheduler is just as slow as on the client—but it's even worse to do it on the scheduler, because then you block the event loop for tens of seconds at a time! If you have tasks already submitted to the scheduler, and then you submit another big graph to the scheduler, things can go very sideways as you essentially take the scheduler offline for half a minute to materialize your graph.

----

What about this incremental approach:
1. Materialize the whole graph on the client. Stream it task-by-task, in priority order (depth-first), to the scheduler. Since it's in priority order, the scheduler can start scheduling tasks as soon as it gets them, without waiting to receive every task. Total compute time is about the same as today (materialize is the same, transfer is slower but now overlaps with compute), but a) you're no longer blocking the scheduler's event loop, improving stability and b) HLG serialization and unpicking on the scheduler becomes a non-issue, since it all happens on the client, so developing HLGs becomes vastly easier.
2. Implement backpressure from the scheduler to the client to say "hey, I have enough (processing) tasks already to saturate my workers (plus a bit of a buffer to hide my latency talking to you), don't send me any more". This would effectively fix root-task overproduction without implementing https://github.com/dask/distributed/issues/3974, because the scheduler wouldn't even be aware of the extra root tasks, therefore it couldn't send them to workers. (You probably do still want STA, combined with this, in the long run though.) With task backpressure, you also wouldn't have to have every task in scheduler memory the whole time (task count would increase linearly until the very end). There are big questions here about whether this would make scheduling too "dumb"—what happens when some tasks are completing faster than others, but the priority order is just sending along dependents of the slow tasks? Would there be some protocol for the scheduler to request dependents of certain keys, rather than blind priority order? (This would probably add way too much latency.)
2. Stream results _back_ from the cluster to the client https://github.com/dask/distributed/issues/4754. Now, we can stream both into and out of the scheduler. As soon as an output key is complete and handed back to the client, the scheduler can forget about it. So the number of tasks in play on the scheduler is usually `Θ(nthreads)`, regardless of the total graph size. This is kinda https://github.com/dask/distributed/issues/5630, except you're "spilling" the unneeded tasks to the client instead of disk!
2. Make it possible to iterate through HLGs in priority order without materialization. Now, when you stream your graph from the client to the scheduler, a) whole graph is never materialized _anywhere_ (client or scheduler), b) work starts nearly instantly, since you don't have to wait for full materialization to start sending the scheduler tasks, c) graph size becomes effectively unlimited for embarrassingly-parallel workloads; the practical limits are the number of workers, and the size of the non-embarrassingly-parallel substructures within the graph.
3. At this point, things seem pretty great to me. Compute starts instantly, you don't have limitations on graph size, and scheduling is better without root-task overproduction. But if you want even more, _maybe_ you move HLG unpacking back to the scheduler, and iterate through them there. Now, HLGs are essentially a compression format as well. And you have to deal with serialization again. But a) the client connection is less critical to scheduling, and b) the scheduler could pick which parts of the HLG to unpack next without client<->scheduler latency (if the interface even supported this).

I'll note that with these ideas, the client becomes an increasingly-integral part of the system, since the cluster no longer holds the full graph, nor all of the results, in memory. That flies in the face of my concerns in https://github.com/dask/distributed/issues/5684. I don't like being increasingly dependent on a fast, stable, long-lived connection to the client, but:
1. The benefits seem significant enough, it's probably worth it.
1. You could use `afar`-like patterns to turn one of the workers into your client, or maybe even spin up a sidecar Python process on the scheduler, etc., to improve resilience.

cc @fjetter @crusaderky

**Comment by rjzamora** [comment]
>In order to make this iterate-in-priority-order concept work—which is where I think we'd really see HLGs pay off—we'd have to go all-in on flipping these graph-generation algorithms on their heads.

Yes - I get the feeling that the most-promising solution is to take the time to flip all common graph-construction algorithms on their heads, or develop a good expression system to somehow simplify/automate this (not that I have such a HLE solution in my head).  

Regarding your graph streaming suggestion/plan:  This approach is certainly in line with what I was thinking. Perhaps we could also start by looping over partition/chunk-specific subgraphs on the client (to avoid the need to materialize the entire graph at once). This is similar to something we do in NVTabular to ensure the entire subgraph for a specific output key is completely executed on the same worker.

**Comment by fjetter** [comment]
> I'll note that with these ideas, the client becomes an increasingly-integral part of the system, since the cluster no longer holds the full graph, nor all of the results, in memory.

Our global state machine is complex and complicated enough as it is without adding a third distinct actor to the mix.
If we want to start streaming the tasks in priority order, that may be something to consider. We'd still need to fully materialize the graph but I would suggest to not couple too many problems at once and would defer this to a later discussion. Implementing some kind of backpressure/feedback mechanism that involves the Client directly or indirectly in scheduling decisions is something I am very strongly against.

**Comment by mrocklin** [comment]
My gut reaction to this is skepticism.  Large graphs are less of a problem
than optimal scheduling.  This approach helps with large graphs at the cost
of optimal scheduling.  I think that the difficulties around HLGs are due
to poor design so far, and that can be fixed.  HLGs also provide a variety
of other long-term benefits to the scheduler, which would have a lot more
semantic information about users' computations, which I think will be
valuable in the future.

I think that trying to schedule things iteratively is a cool idea, but is
likely to be even more complex than high level graphs.

**Comment by rjzamora** [comment]
Thanks for the feedback @mrocklin!  Note that I have been completely on the fence about this idea from the start (and I still very-much am) :) Overall, I agree with everything you said.  With that in mind…

>Large graphs are less of a problem than optimal scheduling.

Totally agree. However, the only reason we are materializing HLG layers on the scheduler is because of graph communication time (which only seems to matter for large graphs and/or remote client-scheduler connections).  I’m okay with saying graph size isn’t a real concern, but then it seems strange that we are taking the time to struggle with multi-step serialization to ultimately minimize the graph-communication cost.

>This approach helps with large graphs at the cost of optimal scheduling.

Yes - A naive layer-by-layer approach would likely inhibit optimal scheduling, but that suggestion was more of a “test” (not how we would ever communicate/execute in practice).  However, I would’t expect the long-term procedure to hurt scheduling at all.  The core idea is just that we allow the client to send an incomplete graph to the scheduler.  Ideally, that subgraph would correspond to tasks that should be scheduler before the following subgraph anyway.

>I think that the difficulties around HLGs are due to poor design so far, and that can be fixed.

Agree that HLGs should be revised, but are you focusing on pack/unpack (serialization) design, or something else?

>HLGs also provide a variety of other long-term benefits to the scheduler, which would have a lot more
semantic information about users' computations, which I think will be valuable in the future. 
I agree that we can get a lot out of HLG layers, but I am also a bit skeptical that we really need to avoid materializing the low-level graph representation (within the Layer) before communicating it to the scheduler.  For example, the annotations that we currently ship with the layer have nothing to do with whether that layer is materialized or not.

**Comment by crusaderky** [comment]
I really like @gjoseph92's idea.
I think there are a few issues with it though:

1. as pointed out already, Client<->Scheduler connectivity is frequently less reliable/predictable than Scheduler<->Worker
2. Client<->Scheduler bandwidth and latency can be much, much slower than the one inside the cluster - namely when the Client is on somebody's laptop and connects to a scheduler on AWS. Worst case scenario: the laptop is connecting over 5G or a Starbucks WiFi.
3. This means that a client cannot be closed anymore after ``publish_dataset``, or that alternatively ``publish_dataset`` must eagerly send to the scheduler the remainder of the graph
4. The Client may hold its own event loop precious too, and this design slams it down hard. This is not the typical case of Jupyter Notebooks, but it is that of e.g. a web or RPC server that uses dask to offload computation.

I would like to propose an improvement on top of @gjoseph92's design: a new Server between the Client and the Scheduler, which I'll call PreScheduler.

1. Each Client connects to a PreScheduler and uploads whole, unmaterialized, unoptimized, HLGs. 
2. The PreScheduler materializes and optimizes the graph
3. After this point, everything is as @gjoseph92 described above: the PreScheduler internally defines order and sends keys to the scheduler on a as-needed basis

- This whole new service is *optional* - it is straightforward to flick a switch to run everything it does inside either the Scheduler or the Client process/event loop, minus either one of the network connections
- The PreScheduler runs near the Scheduler (e.g. on AWS).
- You can have as many PreSchedulers as there are Clients.
- You can write adaptive logic that starts and stops on demand new PreSchedulers on Kubernetes/other depending on the amount of Clients connected
- You can also have a simpler system that spawns a new local process on the Scheduler host every time a new client connects.

The connection process goes as follows:
1.  1a. A pool of statically started ``dask-prescheduler`` processes registers to the Scheduler, exactly like ``dask-worker`` does, OR
    1b. Scheduler starts a fixed-size pool of subprocesses; once there are more Clients than subprocesses, more than one Client will start connecting to the same PreScheduler, OR
    1c. Scheduler has the capability to spawn and destroy cloud containers running ``dask-prescheduler`` on demand
2. Client connects to Scheduler
3. The Scheduler *might* start a new process or cloud container; see point 1. This is invisible to the Client.
4. Scheduler informs Client of the IP:port of the PreScheduler
5. Client connects to PreScheduler
6. After the Client disconnects, wait a certain timeout (in case it reconnects) and then shut down the PreScheduler
